File: Sources/VideoEditCore/BoundaryScore.swift
import Foundation

private let weakConnectives: Set<String> = [
    "e", "ma", "o", "oppure", "che", "perché", "perche", "quindi",
    "allora", "mentre", "invece", "comunque", "però", "pero",
    "dunque", "infatti", "cioè", "cioe", "inoltre",
    "anche", "poi", "tuttavia", "anzi",
]

private let strongStarts: Set<String> = [
    "no", "sì", "si", "ah", "oh", "beh", "ecco", "ok", "okay",
    "bene", "giusto", "certo",
]

private let pauseWords: Set<String> = [
    "diciamo", "praticamente", "fondamentalmente", "sostanzialmente",
]

public struct BoundaryContext: Sendable {
    public var prevGap: Double
    public var silenceAfter: Double

    public init(prevGap: Double = 0, silenceAfter: Double = 0) {
        self.prevGap = prevGap
        self.silenceAfter = silenceAfter
    }
}

public enum BoundaryScore {
    public static func evaluate(
        currText: String,
        currEnd: Double,
        nextText: String,
        nextStart: Double,
        profile: VideoProfile,
        context: BoundaryContext? = nil
    ) -> Float {
        let curr = currText.trimmingCharacters(in: .whitespaces)
        let nxt = nextText.trimmingCharacters(in: .whitespaces)

        guard !curr.isEmpty, !nxt.isEmpty else { return 1.0 }

        let gapSec = nextStart - currEnd
        var score: Float = 0

        let gap = Float(min(gapSec / profile.gapBreak, 1.5))
        score += gap * 0.25

        if curr.count > profile.maxChars {
            score += 0.4
        }

        let endsStrong = curr.last.map { ".!?…".contains($0) } ?? false
        let endsComma = curr.last.map { ",;:".contains($0) } ?? false

        if endsStrong {
            score += 0.7
        } else if endsComma {
            score -= 0.4
        }

        if let firstChar = nxt.first {
            if firstChar.isUppercase {
                score += 0.3
            } else {
                score -= 0.3
            }
        }

        let firstWord = nxt.split(separator: " ").first.map {
            $0.lowercased().trimmingCharacters(in: CharacterSet(charactersIn: "«»\"'"))
        } ?? ""

        if weakConnectives.contains(firstWord) {
            if !endsStrong {
                score -= 0.5 * profile.weakConjBoost
            }
        }
        if strongStarts.contains(firstWord) {
            score += 0.3
        }
        if pauseWords.contains(firstWord) {
            score -= 0.3
        }

        let currWords = curr.split(separator: " ").count
        let lenPenalty: Float
        if currWords <= 2 { lenPenalty = 0.5 }
        else if currWords <= 4 { lenPenalty = 0.2 }
        else { lenPenalty = 0 }

        if !endsStrong { score -= lenPenalty }

        if let ctx = context {
            if ctx.prevGap > 0 && gapSec < ctx.prevGap * 0.3 {
                score -= 0.2
            }
            if ctx.silenceAfter > 2.0 {
                score += 0.3
            }
        }

        return score
    }

    public static func shouldBreak(
        currText: String,
        currEnd: Double,
        nextText: String,
        nextStart: Double,
        profile: VideoProfile,
        context: BoundaryContext? = nil
    ) -> Bool {
        evaluate(
            currText: currText, currEnd: currEnd,
            nextText: nextText, nextStart: nextStart,
            profile: profile, context: context
        ) >= profile.boundaryThreshold
    }
}

File: core/boundary_score.py
import re

_WEAK_CONNECTIVES = {
    "e", "ma", "o", "oppure", "che", "perché", "perche", "quindi",
    "allora", "mentre", "invece", "comunque", "però", "pero",
    "dunque", "quindi", "infatti", "cioè", "cioe", "inoltre",
    "anche", "poi", "tuttavia", "anzi",
}

_STRONG_STARTS = {
    "no", "sì", "si", "ah", "oh", "beh", "ecco", "ok", "okay",
    "allora", "dunque", "bene", "giusto", "certo",
}

_PAUSE_WORDS = {
    "diciamo", "praticamente", "fondamentalmente", "sostanzialmente",
}


def boundary_score(
    curr_text: str,
    curr_end: float,
    next_text: str,
    next_start: float,
    profile: dict,
    context: dict | None = None,
) -> float:
    curr = curr_text.strip()
    nxt = next_text.strip()
    if not curr or not nxt:
        return 1.0

    gap_sec = next_start - curr_end
    score = 0.0

    gap = min(gap_sec / profile.get("gap_break", 1.0), 1.5)
    score += gap * 0.25

    if len(curr) > profile["max_chars"]:
        score += 0.4

    if curr and curr[-1] in ".!?…":
        score += 0.7
    elif curr and curr[-1] in ",;:":
        score -= 0.4

    if nxt:
        if nxt[0].isupper():
            score += 0.3
        else:
            score -= 0.3

    first_word = nxt.split()[0].lower().strip("«»\"'") if nxt.split() else ""
    ends_strong = curr and curr[-1] in ".!?…"
    if first_word in _WEAK_CONNECTIVES:
        if not ends_strong:
            score -= 0.5 * profile.get("weak_conj_boost", 1.0)
    if first_word in _STRONG_STARTS:
        score += 0.3
    if first_word in _PAUSE_WORDS:
        score -= 0.3

    curr_words = curr.split()
    len_penalty = 0.0
    if len(curr_words) <= 2:
        len_penalty = 0.5
    elif len(curr_words) <= 4:
        len_penalty = 0.2

    if curr and curr[-1] not in ".!?…":
        score -= len_penalty

    if context:
        prev_gap = context.get("prev_gap", 0)
        if prev_gap > 0 and gap_sec < prev_gap * 0.3:
            score -= 0.2

        silence_after = context.get("silence_after", 0)
        if silence_after > 2.0:
            score += 0.3

    return score


def should_break(
    curr_text: str,
    curr_end: float,
    next_text: str,
    next_start: float,
    profile: dict,
    context: dict | None = None,
) -> bool:
    return boundary_score(curr_text, curr_end, next_text, next_start, profile, context) >= profile["boundary_threshold"]
